### PR TITLE
Mosek: Draft for handling time limits

### PR DESCRIPTION
**Pull request by Paulnkk**
## Description
I introduced conditions that are triggered if a max time limit is reached and throws a user warning.

Issue link (if applicable):
https://github.com/cvxpy/cvxpy/issues/1158

## Type of change
- [x] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [ ] Bug fix
- [ ] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [ ] Add our license to new files.
- [ ] Check that your code adheres to our coding style.
- [ ] Write unittests.
- [x] Run the unittests and check that they’re passing.
- [ ] Run the benchmarks to make sure your change doesn’t introduce a regression.

### CODE REVIEW

**Comment by Paulnkk** [comment]
If the time limit triggers and a solver error is raised, the user can at least still see the best solution generated by mosek. 

Currently the question is whether a return as specified above is desired. Furthermore I will add unit tests and fix conflicts with the existing ones.

**Comment by Paulnkk** [comment]
@aszekMosek Hey, I opened this Mosek-related PR and wanted to pick it up again. Does this PR generally satisfy the purpose of the issue mentioned above (https://github.com/cvxpy/cvxpy/issues/1158)?

**Comment by aszekMosek** [comment]
If the solver terminated because of time limit then the return value of ``task.optimize()`` will be ``rescode.trm_max_time ``  (https://docs.mosek.com/latest/pythonapi/optimizer-task.html#mosek.task.optimize). I think that will be a much more simple and reliable way to check for this situation, rather than the more complicated logic involving prosta etc., which can be much more involved and depend on which solver is used, etc.

**Comment by aszekMosek** [comment]
On second thought, I don't really understand what is going on here: if the solution status is ``unknown`` then CVXPY normally returns that the solver failed. Why should it return something else if the solver hit the time limit? How is that different than any other ``unknown`` situation?

If the solver thinks it has something interesting, then it will return some other status, and then CVXPY can make the solution available.

If the solver has nothing, then it has nothing, regardless if it is because of time limit or anything else.

The only thing I think should be done is that if the termination code is ``rescode.trm_max_time`` then set some flag/status informing the user about it, if possible. 

Note that there already was a PR recently which allows the user to say that they want to fetch anything, regardless of solution status being ``unknown``.

**Comment by Paulnkk** [comment]
@aszekMosek, thanks for your suggestions, and sorry for getting back on the PR that late. I inserted a condition to throw a user warning if the time limit is reached within the solver (please let me know if the user warning satisfies what you mentioned as a flag). Benchmarks changes which may be related to the case that `task.optimize()` is called a second time.

File: cvxpy/reductions/solvers/conic_solvers/mosek_conif.py [hunk 1]
```diff
@@ -296,6 +296,11 @@ def solve_via_data(self, data, warm_start: bool, verbose: bool, solver_opts, sol
         # Optimize the Mosek Task, and return the result.
         task.optimize()
 
+        if task.optimize() == mosek.rescode.trm_max_time:
+            warnings.warn(
+                "Optimization terminated by time limit; solution may be imprecise or absent.",
+            )
+
```
Comment (phschiele) [inline cvxpy/reductions/solvers/conic_solvers/mosek_conif.py:303]:
```suggestion
        rescode = task.optimize()

        if rescode == mosek.rescode.trm_max_time:
            warnings.warn(
                "Optimization terminated by time limit; solution may be imprecise or absent.",
            )

```

**Review by phschiele — COMMENTED** [review]
There is an empty `test.py` file floating around. Made a suggestion to only call optimize once.

**Comment by Paulnkk** [comment]
> There is an empty `test.py` file floating around. Made a suggestion to only call optimize once.

@phschiele Removed test file here, leftover from debugging ...

**Review by phschiele — APPROVED** [review]
Thanks for the swift update @Paulnkk!

**Review by SteveDiamond — APPROVED** [review]

**Pull request merged**